### PR TITLE
feat: conversational memory system with 4-layer architecture + 13 storage adapters

### LINKED ISSUE #215: implement human-like memory system for agents
### what this is about

agents need better memory than just stuffing everything in context windows. been seeing too many failures in production when conversations go beyond a few hours.

this implements an advanced memory system based on how humans actually remember things - patterns that work in real deployments.

### core system needed

**4-layer memory architecture:**
- working memory: current conversation context with ttl management
- episodic memory: timestamped experiences that decay over time  
- semantic memory: facts extracted from repeated patterns
- procedural memory: tool usage patterns that actually worked

**batch processing system:**
- message buffering for real-world usage patterns
- 5x cost reduction through extraction rate control
- three-tier extraction (rules → small llm → large llm)
- noise filtering and deduplication
- batch metadata tracking

**unified memory search:**
- cross-memory type search and retrieval
- intelligent relevance scoring and ranking
- hybrid search (text + vector + temporal)
- memory relationship discovery
- configurable search strategies

**lifecycle management:**
- automated decay with configurable rules
- memory promotion between types (episodic → semantic)
- cleanup and archival operations
- memory limit enforcement
- evolution tracking

**intelligence layer:**
- connection detection between memories
- semantic linking with embedding support
- configurable rule-based triggers
- optional llm enhancement (cost-controlled)

### technical requirements

**storage foundation:**
- vector search capabilities (sqlite-vec, postgresql-vector, pinecone, qdrant)
- namespace isolation for multi-agent systems
- ttl support for session management  
- batch operations for memory loading

**memory system implementation:**
- memory manager orchestrating all 4 types
- configurable decay engine with access boost algorithms
- batch processor with cost optimization
- recall service for unified search
- lifecycle manager for automation

### why this matters

current approaches fail because:
1. context windows fill up with irrelevant data
2. important info gets pushed out by recent chatter
3. no learning from what actually works for users
4. every conversation starts from zero

this fixes those issues by:
- only remembering what matters (rule-based + ai triggers)
- converting old experiences into lasting knowledge
- learning from successful tool usage patterns
- maintaining critical info with configurable decay
- building semantic networks of interconnected memories

### success metrics

- maintain conversation coherence beyond 300 turns
- reduce memory-related failures by 90%
- capture and reuse successful tool patterns
- support configurable retention rules per use case
- demonstrate automated memory promotion and cleanup

---

### future enhancements (lower priority)

**additional memory types:**
- preference memory: user/agent preferences tracking
- relationship memory: connections between entities

**advanced evolution:**
- dynamic memory reorganization engine
- automatic rule learning from conversation patterns
- cross-conversation memory sharing
- memory network visualization

**optimization features:**
- advanced semantic linking algorithms
- memory importance learning
- advanced analytics and insights dashboard

this is practical advanced memory management for production agents that need sophisticated memory capabilities, learning from successes, and maintaining reliability over extended use.

### PR DESCRIPTION
**Pull request by oguzserdar**
built the complete cognitive memory architecture for agentdock - this is the brain that makes agents truly intelligent. includes both the storage foundation (13 adapters) and the full 4-layer memory system on top.

## memory system (the main event)

**4-layer cognitive architecture:**

- working memory - current conversation context with ttl management
- episodic memory - conversation history with semantic search
- semantic memory - knowledge extraction and long-term facts
- procedural memory - learned patterns and agent workflows

**production-grade prime extraction:**

- prime extractor with intelligent tier selection (tested with real scenarios)
- recallservice for unified cross-memory search with hybrid search
- memorymanager for system orchestration
- intelligence layer with connections and consolidation

**real performance wins:**

- memory operations complete in &lt;100ms (non-blocking)
- prime system reduces extraction costs by 80% (batch processing)
- hybrid search with 30% text + 70% vector fusion
- production-ready with proper decay and lifecycle management

## storage foundation (13 adapters)

**local to production:**

- sqlite (default) - zero config, works everywhere
- sqlite-vec - sqlite + vector search for semantic memory
- postgresql - production db with proper pooling + memory tables
- postgresql-vector - pgvector for embeddings storage

**cloud natives:**

- s3 - large context storage
- dynamodb, cloudflare kv/d1 - serverless options
- redis, vercel kv - existing ones refactored

**vector specialists:**

- pinecone, qdrant, chromadb - dedicated vector dbs
- memory provider - in-memory for testing

**browser/server separation:**

- proper dual exports following industry standards (prisma/nextjs pattern)
- browser-safe main exports (@agentdock/core)
- server-only exports for node.js adapters (@agentdock/core/server)
- no more "can't resolve fs" errors in nextjs

## agent memory patterns

```typescript
// memory system just works
const memory = await createMemorySystem({
  environment: 'production',
  recallPreset: 'precision'
});

// store conversation context
await memory.store(userId, agentId, "user asked about pricing");

// recall with hybrid search across all memory types
const memories = await memory.recall(userId, agentId, "pricing questions");

// prime extraction for cost optimization
const extracted = await primeExtractor.extract(messages, {
  tier: 'balanced',
  costOptimization: true
});
```

## what's tested and working

**memory system:**

- comprehensive test suite with 47/47 suites passing
- real-world production data validation
- cost optimization verified (80% reduction achieved // with prime)
- performance targets met (&lt;100ms operations // on local)
- hybrid search algorithms validated // 30-70 split

**storage adapters:**

- production validation across all 13 adapters
- namespace isolation verified
- ttl behavior tested
- vector operations optimized
- concurrent access patterns validated

**integration:**

- nextjs build working with browser/server separation
- typescript compilation clean
- type system unified with security fixes
- documentation restructured with proper cross-linking

## production architecture

this gives agents proper memory that:

- remembers past conversations (episodic)
- learns domain knowledge (semantic)
- improves workflows over time (procedural)
- maintains context efficiently (working)
- scales with proper storage backends
- optimizes costs through prime extraction
- searches intelligently with hybrid fusion

the storage layer handles all persistence. the memory layer makes agents intelligent.

closes #215

## Summary by CodeRabbit

- **New Features**

  - Introduced a comprehensive, multi-layered memory system with working, episodic, semantic, and procedural memories.
  - Added unified recall service with hybrid vector/text search, caching, and advanced relevance scoring.
  - Implemented lazy decay system with batch processing for efficient memory updates.
  - Delivered PRIME intelligent memory extraction system and orchestrator for streamlined memory ingestion.
  - Added memory connection management, consolidation, temporal pattern analysis, and cost tracking.
  - Provided production-grade encryption service for sensitive memory data.
  - Included extensive configuration presets and example agent templates for therapy, business, and research domains.
  - Added new storage adapters and enhanced existing ones, including vector database integrations (Pinecone, Qdrant, ChromaDB), cloud storage (S3, DynamoDB, Cloudflare KV/D1), and improved PostgreSQL and MongoDB support.
  - Introduced memory lifecycle management with configurable decay, promotion, and cleanup policies.
  - Added server-only exports for Node.js-specific storage adapters.
  - Provided detailed environment configuration expansions with new storage and memory options.
  - Added embedding model creation utilities supporting multiple AI providers.
  - Introduced comprehensive connection graph and temporal pattern analysis tools.
  - Added utilities and base classes for memory types with automatic connection discovery and pattern analysis.
  - Provided extensive utility functions for semantic, episodic, procedural, and working memory data management.
  - Added memory evolution tracking for analytics.
  - Introduced cost tracking system for memory extraction operations.
  - Added comprehensive storage adapter registration system with dynamic imports and optional Node.js-only adapters.
  - Added new test utilities and mock providers for embedding and memory connection testing.

- **Bug Fixes**

  - Improved validation and error handling across memory, recall, and storage operations.
  - Fixed strict user and agent isolation enforcement in all memory operations.

- **Documentation**

  - Added detailed developer-focused documentation covering memory system architecture, storage adapters, lifecycle management, intelligence layer, and extraction system.
  - Provided usage examples, configuration guides, and best practices for all components.
  - Added new README files for storage adapters and memory subsystems.

- **Tests**

  - Added extensive unit, integration, end-to-end, and performance tests covering memory types, lazy decay, recall service, PRIME extraction, and storage adapters.
  - Included new test suites for memory manager, lazy decay components, memory consolidator, connection system, and storage adapter factories.

- **Chores**

  - Updated environment configuration examples and dependency management for better local development experience.
  - Streamlined package development workflow with new development guide.

## Summary by CodeRabbit

* **New Features**
  * Introduced a comprehensive, multi-layered memory system with working, episodic, semantic, and procedural memory types, including support for lazy decay, hybrid search, memory consolidation, connection discovery, and cost tracking.
  * Added advanced storage adapters for PostgreSQL, SQLite, MongoDB, DynamoDB, S3, Cloudflare KV/D1, Pinecone, Qdrant, and ChromaDB, with unified interfaces and batch operations.
  * Implemented an intelligence layer with embedding services, connection management, temporal pattern analysis, and LLM-based extraction and consolidation.
  * Provided production-ready configuration presets for local and production environments, plus agent-specific lifecycle templates for therapy, business, and research use cases.
  * Added robust encryption, recall, and memory evolution tracking services.
  * Extensive documentation and usage guides for all storage, memory, and intelligence modules.
  * Introduced PRIME memory extraction system with embedded rule guidance and tiered model selection.
  * Added MemoryConnectionManager for progressive, cost-optimized memory relationship discovery.
  * Added MemoryConsolidator for semantic memory synthesis and hierarchical abstraction.
  * Added EmbeddingService with caching and batch processing for vector generation.
  * Added RecallService for unified multi-memory search with hybrid scoring and caching.
  * Introduced LazyDecayCalculator and LazyDecayBatchProcessor for on-demand memory decay.
  * Added comprehensive storage adapter registration system with dynamic imports.
  * Introduced detailed lifecycle management with decay, promotion, and cleanup configurations.
  * Added detailed test utilities and mock implementations for memory and connection testing.
  * Added server-only exports for Node.js-specific storage adapters.

* **Bug Fixes**
  * Improved validation of query parameters and memory configuration, ensuring correct handling of boundary values and error scenarios.

* **Tests**
  * Added comprehensive unit, integration, and end-to-end test suites covering memory system workflows, storage adapters, lazy decay, connection discovery, recall, consolidation, and performance.
  * Added performance tests for WorkingMemory store latency and lazy decay batch processing.
  * Added integration tests for memory consolidator, connection graph, connection-enhanced recall, and temporal pattern analysis.
  * Added end-to-end tests for lazy decay system and memory connection system.

* **Documentation**
  * Added detailed README files, configuration guides, and adapter-specific documentation for storage, memory, and intelligence modules.
  * Provided extensive developer-focused memory system and intelligence layer documentation.
  * Added example configurations for therapy, business, and research agents with lifecycle and procedural memory settings.
  * Added development guide for agentdock-core package setup and workflow.

* **Chores**
  * Updated dependency versions, added type declarations for external SDKs, and improved configuration templates for environment variables and development workflows.
  * Added Jest configuration for test-specific TypeScript settings.
  * Refined environment variable examples with expanded storage and memory configuration options.
  * Reorganized exports and module entry points for clarity and modularity.

* **Style**
  * Refactored code for clarity, maintainability, and modularity, with enhanced logging and error reporting.
  * Improved error handling with detailed messages and fallback mechanisms across memory and storage layers.

### CODE REVIEW

**Comment by oguzserdar** [comment]
hey @coderabbitai can you review the recent storage optimizations in commit 362286e? we implemented 10 out of your 92 original suggestions from this pr. we deliberately ignored about 82 of your comments because our validation showed they were either false positives (like suggesting fixes for code that was already working correctly) or low-value nitpicks (style preferences, over-engineering suggestions, edge cases that rarely matter). want to see if you still think those ignored items are worth fixing or if you agree with our filtering approach. particularly curious about your take on the memory-first dependency architecture we went with.

**Comment by oguzserdar** [comment]
@coderabbit full review, and consider this we implemented 25 critical fixes from your review instead of all 92 

focused on security vulnerabilities, performance bottlenecks, and actual bugs. remaining 67 items are style preferences and edge cases deemed non-critical for production. achieved 99%ish test success rate with all security issues resolved. please review again.

File: agentdock-core/src/memory/intelligence/consolidation/MemoryConsolidator.ts [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+    if (!userId?.trim()) {
+      throw new Error('userId is required for memory retrieval operations');
+    }
+
+    // Use storage memory operations if available
+    if (this.storage.memory?.getByType) {
+      return this.storage.memory.getByType(userId, agentId, 'episodic', {
+        createdBefore: cutoffTime
+      });
+    }
```
[the remark this answers is not in the document]
Comment (oguzserdar) [reply agentdock-core/src/memory/intelligence/consolidation/MemoryConsolidator.ts:604]: Code properly uses this.storage.memory?.getByType with fallback
Has proper warning logs when unavailable - not a bug

**Comment by oguzserdar** [comment]
@coderabbitai full review please!

implemented fixes for the legitimate functional gaps found in pr 222 review

verified each issue through direct code examination and fixed 4 real problems while identifying 6 false positives

fixed issues:
- decay engine constructor missing config validation causing potential runtime crashes
- connection graph placeholder methods returning empty arrays breaking graph insights
- memory consolidator similarity detection just returning target memory instead of finding similar ones

implemented clean solutions without backward compatibility baggage since this is feature branch:
- strict config validation with clear error messages and fail fast approach
- proper connection retrieval using standard storage patterns
- connected components clustering algorithm with dfs traversal
- multi method similarity detection with vector search primary and content fallback

removed unnecessary fallback patterns and defensive error handling to make code more direct and maintainable

false positives were non existent console.log statements and already complete interfaces

File: agentdock-core/DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+When `agentdock-core` is published:
+```json
+{
+  "dependencies": {
+    "agentdock-core": "^1.0.0"
+  }
+}
+```
+
+The postinstall complexity will be removed entirely.
```
[the remark this answers is not in the document]
Comment (oguzserdar) [reply agentdock-core/DEVELOPMENT.md:48]: we're future proofing here, no worries.

**Comment by oguzserdar** [comment]
@coderabbitai full review, per favore
